oracle/oracle.go

```diff
@@ -609,3 +609,117 @@ func (o *oracleOps) GetDeviceID(vol interface{}) (string, error) {
 	}
 	return "", fmt.Errorf("invalid type: %v given to GetDeviceID", vol)
 }
+
+// Enumerate volumes that match given filters. Organize them into
+// sets identified by setIdentifier.
+// labels can be nil, setIdentifier can be empty string.
+func (o *oracleOps) Enumerate(volumeIds []*string,
+	labels map[string]string,
+	setIdentifier string,
+) (map[string][]interface{}, error) {
+	sets := make(map[string][]interface{})
+	req := core.ListVolumesRequest{
+		CompartmentId: common.String(o.compartmentID),
+	}
+	resp, err := o.storage.ListVolumes(context.Background(), req)
+	if err != nil {
+		return nil, err
+	}
+	volIDsMap := map[string]string{}
+	for _, volIds := range volumeIds {
+		volIDsMap[*volIds] = *volIds
+	}
+	for _, vol := range resp.Items {
+		_, ok := volIDsMap[*vol.Id]
+		if !ok {
+			continue
+		}
+
+		if o.deleted(vol) {
+			continue
+		}
+		// TODO: [PWX-26616] Check if SDK itself returns list of volumes
+		// that have labels OR use volumeGroup for filtering
+		if labels != nil && containsMap(vol.FreeformTags, labels) {
+			continue
+		}
+		if len(setIdentifier) == 0 {
+			cloudops.AddElementToMap(sets, vol, cloudops.SetIdentifierNone)
+		} else {
+			found := false
+			for tagKey, tagValue := range vol.FreeformTags {
+				if tagKey == setIdentifier {
+					cloudops.AddElementToMap(sets, vol, tagValue)
+					found = true
+					break
+				}
+			}
+			if !found {
+				cloudops.AddElementToMap(sets, vol, cloudops.SetIdentifierNone)
+			}
+		}
+	}
+	return sets, nil
+}
+
+func containsMap(mainMap map[string]string, subMap map[string]string) bool {
+	for k, v := range subMap {
+		value, ok := mainMap[k]
+		if !ok {
+			return false
+		}
+		if value != v {
+			return false
+		}
+	}
+	return true
+}
+
+func (o *oracleOps) deleted(v core.Volume) bool {
+	return v.LifecycleState == core.VolumeLifecycleStateTerminating ||
+		v.LifecycleState == core.VolumeLifecycleStateTerminated
+}
+
+// ApplyTags will apply given labels/tags on the given volume
+func (o *oracleOps) ApplyTags(volumeID string, labels map[string]string) error {
```

Comment:
Yes, it will overwrite the existing tags with newly provided tags. Below is the output of a test exucution which depicts the same:

```
[opc@oke-ccp4cglwkwq-nc7xnngprgq-sslsa6n23jq-0 oracle]$ go test -v .
=== RUN   TestAll
2022/09/20 07:12:03 volume [ocid1.volume.oc1.us-sanjose-1.abzwuljroq26vzoqxg3ofsnosjzi7gfzxajivqisyyocybszi3mdqev2jpea] is still in [PROVISIONING] state Next retry in: 3s
2022/09/20 07:12:06 volume [ocid1.volume.oc1.us-sanjose-1.abzwuljroq26vzoqxg3ofsnosjzi7gfzxajivqisyyocybszi3mdqev2jpea] is still in [PROVISIONING] state Next retry in: 3s
2022/09/20 07:12:09 volume [ocid1.volume.oc1.us-sanjose-1.abzwuljroq26vzoqxg3ofsnosjzi7gfzxajivqisyyocybszi3mdqev2jpea] is still in [PROVISIONING] state Next retry in: 3s
2022/09/20 07:12:12 volume [ocid1.volume.oc1.us-sanjose-1.abzwuljroq26vzoqxg3ofsnosjzi7gfzxajivqisyyocybszi3mdqev2jpea] is still in [PROVISIONING] state Next retry in: 3s
Created disk: { AvailabilityDomain=zSZL:US-SANJOSE-1-AD-1 CompartmentId=ocid1.compartment.oc1..aaaaaaaaxalgtzvncgpshcpiynag6aq5z36vwh63b3aaj63ay3ymkhxgprvq DisplayName=openstorage-test-b8aafea0-5d5d-4eb8-aa73-656afe662b9c Id=ocid1.volume.oc1.us-sanjose-1.abzwuljroq26vzoqxg3ofsnosjzi7gfzxajivqisyyocybszi3mdqev2jpea LifecycleState=AVAILABLE SizeInMBs=51200 TimeCreated=2022-09-20 07:12:03.074 +0000 UTC DefinedTags=map[Oracle-Tags:map[CreatedBy:oracleidentitycloudservice/vshinde CreatedOn:2022-09-20T07:12:03.060Z]] FreeformTags=map[] SystemTags=map[] IsHydrated=true KmsKeyId=<nil> VpusPerGB=10 SizeInGBs=50 SourceDetails=<nil> VolumeGroupId=<nil> IsAutoTuneEnabled=false AutoTunedVpusPerGB=<nil> BlockVolumeReplicas=[] }
=== Initial tags for disk [ocid1.volume.oc1.us-sanjose-1.abzwuljroq26vzoqxg3ofsnosjzi7gfzxajivqisyyocybszi3mdqev2jpea]: ===
=== Updated list tags for disk [ocid1.volume.oc1.us-sanjose-1.abzwuljroq26vzoqxg3ofsnosjzi7gfzxajivqisyyocybszi3mdqev2jpea]: ===
source : openstorage-test 
Test : UPPER_CASE 
foo : bar 
=== Now applying below tags ===
abc : xyz 
=== Updated list tags for disk [ocid1.volume.oc1.us-sanjose-1.abzwuljroq26vzoqxg3ofsnosjzi7gfzxajivqisyyocybszi3mdqev2jpea]: ===
abc : xyz 
--- PASS: TestAll (13.73s)
PASS
ok      github.com/libopenstorage/cloudops/oracle       13.746s
```

The test code looks like below:
```
func tags(t *testing.T, driver cloudops.Ops, diskName string) {
	tags, err := driver.Tags(diskName)
	require.NoError(t, err, "failed to get tags for disk")

	fmt.Printf("=== Initial tags for disk [%s]: ===\n", diskName)
	for key, value := range tags {
		fmt.Printf("%s : %s \n", key, value)
	}
	err = driver.ApplyTags(diskName, diskLabels)
	if _, typeOk := err.(*cloudops.ErrNotSupported); typeOk {
		return
	}

	require.NoError(t, err, "failed to apply tags to disk")

	tags, err = driver.Tags(diskName)
	require.NoError(t, err, "failed to get tags for disk")

	fmt.Printf("=== Updated list tags for disk [%s]: ===\n", diskName)
	for key, value := range tags {
		fmt.Printf("%s : %s \n", key, value)
	}

	diskLabels = map[string]string{"abc": "xyz"}
	fmt.Printf("=== Now applying below tags ===\n")
	for key, value := range diskLabels {
		fmt.Printf("%s : %s \n", key, value)
	}

	err = driver.ApplyTags(diskName, diskLabels)
	if _, typeOk := err.(*cloudops.ErrNotSupported); typeOk {
		return
	}

	tags, err = driver.Tags(diskName)
	require.NoError(t, err, "failed to get tags for disk")

	fmt.Printf("=== Updated list tags for disk [%s]: ===\n", diskName)
	for key, value := range tags {
		fmt.Printf("%s : %s \n", key, value)
	}

	//require.Len(t, tags, 3, "invalid number of labels found on disk")

	err = driver.RemoveTags(diskName, diskLabels)
	require.NoError(t, err, "failed to remove tags from disk")

	tags, err = driver.Tags(diskName)
	require.NoError(t, err, "failed to get tags for disk")
	require.Len(t, tags, 0, "invalid number of labels found on disk")

	err = driver.ApplyTags(diskName, diskLabels)
	require.NoError(t, err, "failed to apply tags to disk")
}
```